server/src/JwtMiddleware.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE OverloadedStrings #-}
+-- | This module provides functionality for verifying the JSON Web Tokens in a wai setting.
+module JwtMiddleware where
+
+import           Control.Applicative
+import           Control.Monad
+import           Data.Aeson            ((.=))
+import qualified Data.Aeson            as Aeson
+import qualified Data.ByteString       as SBS
+import qualified Data.ByteString.Lazy  as LBS
+import qualified Data.List             as List
+import qualified Data.Text             as Text
+import           Data.Time.Clock.POSIX (POSIXTime)
+import qualified Data.Time.Clock.POSIX as Clock
+import qualified Network.HTTP.Types    as Http
+import qualified Network.Wai           as Wai
+import qualified Web.JWT               as JWT
+
+import           AccessControl
+import           JwtAuth
+import           Store                 (Path)
+
+-- | Defines the kinds of errors that cause authorization to fail.
+data AuthError
+  = TokenError TokenError
+    -- ^ Authorization was denied due to an invalid token.
+  | OperationNotAllowed
+    -- ^ Authorization was denied because the operation is not allowed by the token.
+
+-- | Result of checking authorization
+data AuthResult
+  = AuthRejected AuthError
+    -- ^ Authorization was denied because of the specified reason
+  | AuthAccepted
+    -- ^ Authorization was successful
+
+-- * Requests
+
+-- | Check whether accessing the given path with the given mode is authorized by
+-- the token supplied in the request headers or query string (which may not be
+-- present, then failing the check).
+isRequestAuthorized :: Http.RequestHeaders -> Http.Query -> POSIXTime -> Maybe JWT.Secret -> Path -> AccessMode -> AuthResult
+isRequestAuthorized headers query now maybeSecret path mode =
+  case getRequestClaim headers query now maybeSecret of
+    Left err -> AuthRejected (TokenError err)
+    Right claim | isAuthorizedByClaim claim path mode
+                  -> AuthAccepted
+                | otherwise
+                  -> AuthRejected OperationNotAllowed
+
+-- | Extract the JWT claim from the request.
+getRequestClaim :: Http.RequestHeaders -> Http.Query -> POSIXTime -> Maybe JWT.Secret -> Either TokenError IcepeakClaim
+getRequestClaim headers query now maybeSecret =
+  let getTokenBytes = maybe (Left $ VerificationError TokenNotFound) Right (findTokenBytes headers query)
+  in case maybeSecret of
+       -- authorization is enabled, but no secret provided, accept all tokens
+       Nothing     -> getTokenBytes >>= extractClaimUnverified
+       Just secret -> getTokenBytes >>= extractClaim now secret
```

Comment:
Make more clear that the comment only applies to the `Nothing` case. For example by moving it in there, or by moving the `Just` case to above the comment.

Reply:
Fixed!